src/blog/adaptor-frame.md

```diff
@@ -0,0 +1,308 @@
+
+## Arguments adaptor frame
+
+Let’s now call `add42` with fewer and more arguments:
+
+```js
+add42();
+add42(1, 2, 3);
+```
+
+The JS developers between us will know that in the first case, `x` will be assigned `undefined` and the function will return `undefined + 42 = NaN`. In the second case, `x` will be assigned `1` and the function will return `43`, the remaining arguments will be ignored. Note that the caller does not know if that will happen. Even if the caller checks the parameter count, the callee could use the rest parameter or the arguments object to access all the other arguments. Actually, the arguments object can even be accessed outside `add42` in sloppy mode.
+
+If we follow the same steps as before, we will first call the builtin `InterpreterPushArgsThenCall`. It will push the arguments to the stack like so:
+
+![](/_img/adaptor-frame/adaptor-push.svg)
+
+Continuing the same procedure as before, we check if the callee is a function object, get its parameter count and patch the receiver to the global proxy. Eventually we reach `InvokeFunctionCode`.
+
+Here instead of jumping to the `Code` in the callee object. We check that we have a mismatch between argument size and parameter count and jump to `ArgumentsAdaptorTrampoline`.
+
+In this builtin, we build an extra frame, the infamous arguments adaptor frame. Instead of explaining what happens inside the builtin, I will just present you the state of the frame before the builtin calls the callee object’s `Code`. Note that this is a proper `x64 call` (not a `jmp`) and after the execution of the callee we will return to the `ArgumentsAdaptorTrampoline`. This is a contrast with `InvokeFunctionCode` that tailcalls.
+
+![](/_img/adaptor-frame/adaptor-frames.svg)
+
+You can see that we create another frame that copies all the arguments necessary in order to have precisely the parameter count of arguments on top of the callee frame. It creates an interface to the callee function, so that the latter does not need to know the number of arguments. The callee will always be able to access its parameters with the same calculation as before, that is, `[ai] = 2 + parameter_count - i - 1`.
+
+V8 has special builtins that understand the adaptor frame whenever it needs to access the remaining arguments through the rest parameter or the arguments object. They will always need to check the adaptor frame type on top of the callee’s frame and then act accordingly.
+
+As you can see, we solve the argument/register access issue, but we create a lot of complexity. Every builtin that needs to access all the arguments will need to understand and check the existence of the adaptor frame. Not only that, we need to be careful to not access stale and old data. Consider the following changes to `add42`:
+
+```js
+function add42(x) { x += 42; return x; }
```

Comment:
Let's make functions formatted multiline, no need to save space here.

Reply:
Done.